chain-factory/src/factory.rs

```diff
@@ -30,29 +36,104 @@ pub trait FactoryModule {
         min_validators: usize,
         max_validators: usize,
         min_stake: BigUint,
+        chain_name: ManagedBuffer,
         additional_stake_required: MultiValueEncoded<StakeMultiArg<Self::Api>>,
     ) {
         let payment_amount = self.call_value().egld_value().clone_value();
         let deploy_cost = self.deploy_cost().get();
         require!(payment_amount == deploy_cost, "Invalid payment amount");
 
-        let caller = self.blockchain().get_caller();
         let source_address = self.chain_config_template().get();
         let metadata =
             CodeMetadata::PAYABLE_BY_SC | CodeMetadata::UPGRADEABLE | CodeMetadata::READABLE;
+        let args = self.get_deploy_chain_config_args(
+            &min_validators,
+            &max_validators,
+            &min_stake,
+            &additional_stake_required,
+        );
+
+        let sc_address = self
+            .tx()
+            .raw_deploy()
+            .from_source(source_address)
+            .code_metadata(metadata)
+            .arguments_raw(args)
+            .returns(ReturnsNewManagedAddress)
+            .sync_call();
+
+        let chain_id = self.generate_chain_id();
+
+        self.all_deployed_contracts(chain_id.clone())
+            .insert(sc_address);
+
+        let chain_info = ChainInfo {
+            name: chain_name,
+            chain_id,
+        };
+
+        self.chain_info().set(chain_info);
+    }
+
+    #[only_owner]
+    #[endpoint(deployHeaderVerifier)]
+    fn deploy_header_verifier(
+        &self,
+        chain_id: ManagedBuffer,
+        bls_pub_keys: MultiValueEncoded<ManagedBuffer>,
+    ) {
+        let source_address = self.header_verifier_template().get();
+        let metadata =
+            CodeMetadata::PAYABLE_BY_SC | CodeMetadata::UPGRADEABLE | CodeMetadata::READABLE;
+
+        let mut args = ManagedArgBuffer::new();
+        args.push_multi_arg(&bls_pub_keys);
```

Comment:
Should the `bls_pub_keys` be in range (min_validators, max_validators) from chain-config?

Reply:
yes, that's a good point, most probably we'll use a storage mapper here for the range of validators